### PR TITLE
[SPARK-37812][SQL] When deserializing an Orc struct, reuse the result row when possible

### PR DESCRIPTION
**Pull request by bersprockets**
### What changes were proposed in this pull request?

When deserializing an Orc struct, reuse the result row when possible.

When the struct's parent is itself a struct, it is safe to reuse the result row, since some code (an ancestor of the parent, or some downstream processing) has already copied the parent's result row, and therefore our result row.

When the parent is a map or array, it is not safe to reuse the result row, because each element of the parent will point to the same result row. In this case, we must copy the result row before updating the parent.

Note, it is safe to reuse the result row even when the parent is a struct but the grandparent is a map or array, because the parent will copy its result row for each element of the grandparent (and thus copy our result row).

### Why are the changes needed?
This change speeds up deserialization of records containing Orc structs.

| struct size | master | pr    | improvement  |
| ----------- | ------ | ----- | ---------    |
| 10 fields   | 1463   | 296   | 4.94 times
| 100 fields  | 11818  | 2391  | 4.94 times |
| 300 fields  | 38938  | 8061  | 4.8 times |
| 600 fields  | 86735  | 30505 | 2.8 times |

The benchmark results for the PR are included in the PR. The benchmark results for master (with the new benchmark added) are [here](https://issues.apache.org/jira/secure/attachment/13038214/master_results.txt).

### Does this PR introduce _any_ user-facing change?
 No.

### How was this patch tested?

- New unit tests to check correctness (since sometimes the result row is reused, and sometimes it is not).
- New benchmarks.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala [hunk 1]
```diff
@@ -294,6 +294,38 @@ object OrcReadBenchmark extends SqlBasedBenchmark {
     }
   }
 
+  def structBenchmark(values: Int, width: Int): Unit = {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala:297]: Thank you. Could you spin off this benchmark addition into a separate PR, @bersprockets ?
We can merge it first.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala [hunk 2]
```diff
@@ -137,26 +137,38 @@ class OrcDeserializer(
...
-            result.setNullAt(i)
-          } else {
-            fieldConverters(i)(i, value)
-          }
-          i += 1
+        val containerUpdater = updater match {
+          case r: RowUpdater => r
+          case _ =>
+            // If the struct is contained by an array or map, we cannot reuse the same result row.
+            // We must copy the result row before setting it into the array or map
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala:151]: The changes look fine but quick question: do you know why we need to copy in case of an array or map of structs?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala (hunk 1 above)
Comment (bersprockets) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala:297]: Will do.

**Comment by dongjoon-hyun** [comment]
cc @williamhyun , too

**Review by HyukjinKwon — COMMENTED** [review]
@bersprockets, what about something like this?

```diff
diff --git a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala
index 7ab556e330e..ebe0786618d 100644
--- a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala
+++ b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala
@@ -137,32 +137,34 @@ class OrcDeserializer(
         v.changePrecision(precision, scale)
         updater.set(ordinal, v)

-      case st: StructType => (ordinal, value) =>
+      case st: StructType =>
         val result = new SpecificInternalRow(st)
         val fieldUpdater = new RowUpdater(result)
         val fieldConverters = st.map(_.dataType).map { dt =>
           newWriter(dt, fieldUpdater)
         }.toArray
-        val orcStruct = value.asInstanceOf[OrcStruct]

-        var i = 0
-        while (i < st.length) {
-          val value = orcStruct.getFieldValue(i)
-          if (value == null) {
-            result.setNullAt(i)
-          } else {
-            fieldConverters(i)(i, value)
+        (ordinal, value) =>
+          val orcStruct = value.asInstanceOf[OrcStruct]
+
+          var i = 0
+          while (i < st.length) {
+            val value = orcStruct.getFieldValue(i)
+            if (value == null) {
+              result.setNullAt(i)
+            } else {
+              fieldConverters(i)(i, value)
+            }
+            i += 1
           }
-          i += 1
-        }

-        updater.set(ordinal, result)
+          updater.set(ordinal, result)

       case ArrayType(elementType, _) => (ordinal, value) =>
         val orcArray = value.asInstanceOf[OrcList[WritableComparable[_]]]
         val length = orcArray.size()
         val result = createArrayData(elementType, length)
-        val elementUpdater = new ArrayDataUpdater(result)
+        val elementUpdater = new ArrayDataUpdater(result, elementType)
         val elementConverter = newWriter(elementType, elementUpdater)

         var i = 0
@@ -182,10 +184,10 @@ class OrcDeserializer(
         val orcMap = value.asInstanceOf[OrcMap[WritableComparable[_], WritableComparable[_]]]
         val length = orcMap.size()
         val keyArray = createArrayData(keyType, length)
-        val keyUpdater = new ArrayDataUpdater(keyArray)
+        val keyUpdater = new ArrayDataUpdater(keyArray, keyType)
         val keyConverter = newWriter(keyType, keyUpdater)
         val valueArray = createArrayData(valueType, length)
-        val valueUpdater = new ArrayDataUpdater(valueArray)
+        val valueUpdater = new ArrayDataUpdater(valueArray, valueType)
         val valueConverter = newWriter(valueType, valueUpdater)

         var i = 0
@@ -255,9 +257,17 @@ class OrcDeserializer(
     override def setFloat(ordinal: Int, value: Float): Unit = row.setFloat(ordinal, value)
   }

-  final class ArrayDataUpdater(array: ArrayData) extends CatalystDataUpdater {
+  final class ArrayDataUpdater(
+      array: ArrayData, elementType: DataType) extends CatalystDataUpdater {
     override def setNullAt(ordinal: Int): Unit = array.setNullAt(ordinal)
-    override def set(ordinal: Int, value: Any): Unit = array.update(ordinal, value)
+    private val setAny: (Int, Any) => Unit = if (elementType.isInstanceOf[StructType]) {
+      (ordinal: Int, value: Any) => array.update(ordinal, value)
+    } else {
+      (ordinal: Int, value: Any) =>
+        // Copy if element is struct as `RowUpdater` can be reused, see also SPARK-37803.
+        array.update(ordinal, value.asInstanceOf[SpecificInternalRow].copy())
+    }
+    override def set(ordinal: Int, value: Any): Unit = setAny(ordinal, value)

     override def setBoolean(ordinal: Int, value: Boolean): Unit = array.setBoolean(ordinal, value)
     override def setByte(ordinal: Int, value: Byte): Unit = array.setByte(ordinal, value)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala [hunk 3]
```diff
@@ -137,26 +137,38 @@ class OrcDeserializer(
...
+            new CatalystDataUpdater {
+              override def set(ordinal: Int, value: Any) = {
+                updater.set(ordinal, value.asInstanceOf[SpecificInternalRow].copy())
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala:154]: One nitpick question might be that if it's faster to copy instead of creating a new row .. but I guess that's less common to create a map or array of structs (compared to flat rows). Would be great to make sure covering that case in the benchmark

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala (hunk 1 above)
Comment (bersprockets) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala:297]: @dongjoon-hyun 
Created pull request for new benchmarks: #35100. I also added a benchmark for structs contained by arrays (to test whether copying rows is any faster or slower than creating new rows), as requested by @HyukjinKwon

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you, @bersprockets . Please do the following.
1. Rebase to the master
2. File a new JIRA for this improvement and update the PR title.

**Comment by bersprockets** [comment]
Benchmarks are running for the rebased branch. I will update the PR when they are available.

**Comment by dongjoon-hyun** [comment]
The benchmark suite is fixed and merged a few minutes ago. Sorry, but could you rebase this PR once more please?
- https://github.com/apache/spark/commit/4a2ba5b22e84fc79b44604c60320aa5ae679e13a

File: sql/hive/benchmarks/OrcReadBenchmark-results.txt [hunk 4]
```diff
@@ -3,220 +3,220 @@ SQL Single Numeric Column Scan
...
 OpenJDK 64-Bit Server VM 1.8.0_312-b07 on Linux 5.11.0-1022-azure
-Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
+Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
 Single Struct Column Scan with 10 Fields:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-Native ORC MR                                      1126           1149          33          0.9        1073.7       1.0X
-Native ORC Vectorized                              1136           1141           7          0.9        1083.4       1.0X
-Hive built-in ORC                                   589            595           8          1.8         561.4       1.9X
+Native ORC MR                                       363            427          78          2.9         346.2       1.0X
+Native ORC Vectorized                               375            442          88          2.8         357.6       1.0X
```
Comment (c21) [inline sql/hive/benchmarks/OrcReadBenchmark-results.txt:166]: just FYI @bersprockets - for vectorized scan of nested column, I have fixed the benchmark to enable vectorization - https://github.com/apache/spark/commit/4a2ba5b22e84fc79b44604c60320aa5ae679e13a .

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala [hunk 5]
```diff
@@ -844,6 +844,160 @@ abstract class OrcSourceSuite extends OrcSuite with SharedSparkSession {
...
+          |      'a1', 1,
+          |      'a2', 2),
+          |    named_struct(
+          |      'a1', 3,
+          |      'a2', 4)
+          |  ) as col1
+          |""".stripMargin)
+      df.write.orc(file.getCanonicalPath)
+      val df2 = spark.read.orc(file.getCanonicalPath)
+      checkAnswer(df2, df.collect().toSeq)
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:865]: Could you test both vectorized and non-vectorized code path? i.e. enabling and disabling `spark.sql.orc.enableNestedColumnVectorizedReader`? The current test case is working, because vectorized read of struct is disabled, so we are going through the non-vectorized code path (the deserialization code path we are fixing here). I am worried about missing this test, when we decide to enable vectorized reader for nested column in upcoming future Spark release. Thanks.

**Review by c21 — COMMENTED** [review]
Thanks @bersprockets for the fix!

File: sql/hive/benchmarks/OrcReadBenchmark-results.txt [hunk 6]
```diff
@@ -3,220 +3,220 @@ SQL Single Numeric Column Scan
 ================================================================================================
 
 OpenJDK 64-Bit Server VM 1.8.0_312-b07 on Linux 5.11.0-1022-azure
-Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
+Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
```
Comment (LuciferYang) [inline sql/hive/benchmarks/OrcReadBenchmark-results.txt:6]: `OrcReadBenchmark-jdk11-results.txt` and `OrcReadBenchmark-jdk17-results.txt` should be updated at the same time

**Comment by bersprockets** [comment]
I rebased the branch. Now I will:

- Regenerate the benchmarks for Java 8, 11, and 17.
- Update the unit tests as suggested by @LuciferYang

**Comment by dongjoon-hyun** [comment]
Thanks, @bersprockets .

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala [hunk 7]
```diff
@@ -652,9 +652,9 @@ abstract class OrcSourceSuite extends OrcSuite with SharedSparkSession {
         StructField("50",
           StructType(
             StructField("20", StringType) ::
-            StructField("30",
-              StructType(
-                StructField("40", StringType) :: Nil)) :: Nil))))
+              StructField("30",
+                StructType(
+                  StructField("40", StringType) :: Nil)) :: Nil))))
```
Comment (bersprockets) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:657]: Oops. I will fix this weird indentation issue.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @bersprockets and all!
Merged to master for Apache Spark 3.3.

**Comment by HyukjinKwon** [comment]
LGTM2. Wondered if https://github.com/apache/spark/pull/35090#pullrequestreview-843151296 is a better approach but don't feel strongly :-).
